File: solana_toolbox_idl/src/toolbox_idl_type_flat_hydrate.rs
use std::collections::HashMap;
use std::sync::Arc;

use anyhow::anyhow;
use anyhow::Context;
use anyhow::Result;

use crate::toolbox_idl_type_flat::ToolboxIdlTypeFlat;
use crate::toolbox_idl_type_flat::ToolboxIdlTypeFlatEnumVariant;
use crate::toolbox_idl_type_flat::ToolboxIdlTypeFlatFieldNamed;
use crate::toolbox_idl_type_flat::ToolboxIdlTypeFlatFieldUnnamed;
use crate::toolbox_idl_type_flat::ToolboxIdlTypeFlatFields;
use crate::toolbox_idl_type_full::ToolboxIdlTypeFull;
use crate::toolbox_idl_type_full::ToolboxIdlTypeFullEnumVariant;
use crate::toolbox_idl_type_full::ToolboxIdlTypeFullFieldNamed;
use crate::toolbox_idl_type_full::ToolboxIdlTypeFullFieldUnnamed;
use crate::toolbox_idl_type_full::ToolboxIdlTypeFullFields;
use crate::toolbox_idl_typedef::ToolboxIdlTypedef;
use crate::toolbox_idl_utils::idl_map_get_key_or_else;

impl ToolboxIdlTypeFlat {
    pub fn try_hydrate(
        &self,
        generics_by_symbol: &HashMap<String, ToolboxIdlTypeFull>,
        typedefs: &HashMap<String, Arc<ToolboxIdlTypedef>>,
    ) -> Result<ToolboxIdlTypeFull> {
        Ok(match self {
            ToolboxIdlTypeFlat::Defined {
                name,
                generics: generics_flat,
            } => {
                let typedef = idl_map_get_key_or_else(typedefs, name)
                    .context("Hydrate Defined: Lookup")?;
                if generics_flat.len() < typedef.generics.len() {
                    return Err(anyhow!(
                        "Insufficient number of generic parameter for {}: expected: {}, found: {}",
                        typedef.name,
                        typedef.generics.len(),
                        generics_flat.len()
                    ));
                }
                let mut generics_full = vec![];
                for (index, generic_flat) in generics_flat.iter().enumerate() {
                    generics_full.push(
                        generic_flat
                            .try_hydrate(generics_by_symbol, typedefs)
                            .with_context(|| {
                                format!(
                                    "Hydrate Defined: {}, Generic[{}]",
                                    name, index
                                )
                            })?,
                    );
                }
                let mut generics_by_symbol = HashMap::new();
                for (generic_name, generic_full) in
                    typedef.generics.iter().zip(generics_full)
                {
                    generics_by_symbol
                        .insert(generic_name.to_string(), generic_full);
                }
                let type_full = typedef
                    .type_flat
                    .try_hydrate(&generics_by_symbol, typedefs)
                    .with_context(|| {
                        format!("Hydrate Defined: {}, Content", name)
                    })?;
                if let Some(serialization) = &typedef.serialization {
                    if serialization == "bytemuck" {
                        let (pod_alignment, pod_size, pod_content) = type_full
                            .bytemuck_typedef(&typedef.name, &typedef.repr)
                            .with_context(|| {
                                format!("Hydrate Defined: {}, Bytemuck", name)
                            })?;
                        return Ok(ToolboxIdlTypeFull::Pod {
                            alignment: pod_alignment,
                            size: pod_size,
                            content: Box::new(pod_content),
                        });
                    }
                }
                ToolboxIdlTypeFull::Typedef {
                    name: typedef.name.clone(),
                    repr: typedef.repr.clone(),
                    content: Box::new(type_full),
                }
            },
            ToolboxIdlTypeFlat::Generic { symbol } => {
                idl_map_get_key_or_else(generics_by_symbol, symbol)
                    .with_context(|| {
                        format!("Hydrate Generic Lookup: {}", symbol)
                    })?
                    .clone()
            },
            ToolboxIdlTypeFlat::Option {
                prefix,
                content: content_flat,
            } => ToolboxIdlTypeFull::Option {
                prefix: prefix.clone(),
                content: Box::new(
                    content_flat.try_hydrate(generics_by_symbol, typedefs)?,
                ),
            },
            ToolboxIdlTypeFlat::Vec {
                prefix,
                items: items_flat,
            } => ToolboxIdlTypeFull::Vec {
                prefix: prefix.clone(),
                items: Box::new(
                    items_flat.try_hydrate(generics_by_symbol, typedefs)?,
                ),
            },
            ToolboxIdlTypeFlat::Array {
                items: items_flat,
                length: length_flat,
            } => ToolboxIdlTypeFull::Array {
                items: Box::new(
                    items_flat.try_hydrate(generics_by_symbol, typedefs)?,
                ),
                length: usize::try_from(
                    *length_flat
                        .try_hydrate(generics_by_symbol, typedefs)?
                        .as_const_literal()
                        .context("Expected a const literal")?,
                )?,
            },
            ToolboxIdlTypeFlat::Struct {
                fields: fields_flat,
            } => ToolboxIdlTypeFull::Struct {
                fields: fields_flat
                    .try_hydrate(generics_by_symbol, typedefs)?,
            },
            ToolboxIdlTypeFlat::Enum {
                prefix,
                variants: variants_flat,
            } => {
                let mut variants_full = vec![];
                for variant_flat in variants_flat {
                    variants_full.push(
                        variant_flat
                            .try_hydrate(generics_by_symbol, typedefs)?,
                    );
                }
                ToolboxIdlTypeFull::Enum {
                    prefix: prefix.clone(),
                    variants: variants_full,
                }
            },
            ToolboxIdlTypeFlat::Padded {
                before,
                min_size,
                after,
                content: content_flat,
            } => ToolboxIdlTypeFull::Padded {
                before: *before,
                min_size: *min_size,
                after: *after,
                content: Box::new(
                    content_flat.try_hydrate(generics_by_symbol, typedefs)?,
                ),
            },
            ToolboxIdlTypeFlat::Const { literal } => {
                ToolboxIdlTypeFull::Const { literal: *literal }
            },
            ToolboxIdlTypeFlat::Primitive { primitive } => {
                primitive.clone().into()
            },
        })
    }
}

impl ToolboxIdlTypeFlatEnumVariant {
    pub fn try_hydrate(
        &self,
        generics_by_symbol: &HashMap<String, ToolboxIdlTypeFull>,
        typedefs: &HashMap<String, Arc<ToolboxIdlTypedef>>,
    ) -> Result<ToolboxIdlTypeFullEnumVariant> {
        let variant_full_fields = self
            .fields
            .try_hydrate(generics_by_symbol, typedefs)
            .with_context(|| {
                format!("Variant: {}({})", self.name, self.code)
            })?;
        Ok(ToolboxIdlTypeFullEnumVariant {
            name: self.name.to_string(),
            code: self.code,
            fields: variant_full_fields,
        })
    }
}

impl ToolboxIdlTypeFlatFields {
    pub fn try_hydrate(
        &self,
        generics_by_symbol: &HashMap<String, ToolboxIdlTypeFull>,
        typedefs: &HashMap<String, Arc<ToolboxIdlTypedef>>,
    ) -> Result<ToolboxIdlTypeFullFields> {
        Ok(match self {
            ToolboxIdlTypeFlatFields::Named(fields) => {
                let mut fields_full = vec![];
                for field in fields {
                    fields_full.push(
                        field
                            .try_hydrate(generics_by_symbol, typedefs)
                            .with_context(|| {
                                format!("Hydrate Named Field: {}", field.name)
                            })?,
                    );
                }
                ToolboxIdlTypeFullFields::Named(fields_full)
            },
            ToolboxIdlTypeFlatFields::Unnamed(fields) => {
                let mut fields_type_full = vec![];
                for (index, field) in fields.iter().enumerate() {
                    fields_type_full.push(
                        field
                            .try_hydrate(generics_by_symbol, typedefs)
                            .with_context(|| {
                                format!("Hydrate Unnamed Field: {}", index)
                            })?,
                    );
                }
                ToolboxIdlTypeFullFields::Unnamed(fields_type_full)
            },
            ToolboxIdlTypeFlatFields::None => ToolboxIdlTypeFullFields::None,
        })
    }
}

impl ToolboxIdlTypeFlatFieldNamed {
    pub fn try_hydrate(
        &self,
        generics_by_symbol: &HashMap<String, ToolboxIdlTypeFull>,
        typedefs: &HashMap<String, Arc<ToolboxIdlTypedef>>,
    ) -> Result<ToolboxIdlTypeFullFieldNamed> {
        Ok(ToolboxIdlTypeFullFieldNamed {
            name: self.name.to_string(),
            content: self.content.try_hydrate(generics_by_symbol, typedefs)?,
        })
    }
}

impl ToolboxIdlTypeFlatFieldUnnamed {
    pub fn try_hydrate(
        &self,
        generics_by_symbol: &HashMap<String, ToolboxIdlTypeFull>,
        typedefs: &HashMap<String, Arc<ToolboxIdlTypedef>>,
    ) -> Result<ToolboxIdlTypeFullFieldUnnamed> {
        Ok(ToolboxIdlTypeFullFieldUnnamed {
            content: self.content.try_hydrate(generics_by_symbol, typedefs)?,
        })
    }
}

File: solana_toolbox_web3/src/ToolboxIdlTypeFlat.hydrate.ts
import { ToolboxIdlTypedef } from './ToolboxIdlTypedef';
import {
  ToolboxIdlTypeFlat,
  ToolboxIdlTypeFlatArray,
  ToolboxIdlTypeFlatConst,
  ToolboxIdlTypeFlatDefined,
  ToolboxIdlTypeFlatEnum,
  ToolboxIdlTypeFlatFieldNamed,
  ToolboxIdlTypeFlatFields,
  ToolboxIdlTypeFlatFieldUnnamed,
  ToolboxIdlTypeFlatGeneric,
  ToolboxIdlTypeFlatOption,
  ToolboxIdlTypeFlatPadded,
  ToolboxIdlTypeFlatString,
  ToolboxIdlTypeFlatStruct,
  ToolboxIdlTypeFlatVec,
} from './ToolboxIdlTypeFlat';
import {
  ToolboxIdlTypeFull,
  ToolboxIdlTypeFullFields,
} from './ToolboxIdlTypeFull';
import { bytemuck } from './ToolboxIdlTypeFull.bytemuck';
import { ToolboxIdlTypePrimitive } from './ToolboxIdlTypePrimitive';

export function hydrate(
  typeFlat: ToolboxIdlTypeFlat,
  genericsBySymbol: Map<string, ToolboxIdlTypeFull>,
  typedefs: Map<string, ToolboxIdlTypedef>,
): ToolboxIdlTypeFull {
  return typeFlat.traverse(hydrateVisitor, genericsBySymbol, typedefs);
}

let hydrateVisitor = {
  defined: (
    self: ToolboxIdlTypeFlatDefined,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull => {
    let typedef = typedefs.get(self.name);
    if (typedef === undefined) {
      throw new Error(`Could not resolve type named: ${self.name}`);
    }
    if (self.generics.length < typedef.generics.length) {
      throw new Error('Insufficient set of generics');
    }
    let genericsFull = self.generics.map((genericFlat: ToolboxIdlTypeFlat) => {
      return hydrate(genericFlat, genericsBySymbol, typedefs);
    });
    let innerGenericsBySymbol = new Map();
    for (let i = 0; i < typedef.generics.length; i++) {
      innerGenericsBySymbol.set(typedef.generics[i], genericsFull[i]);
    }
    let typeFull = hydrate(typedef.typeFlat, innerGenericsBySymbol, typedefs);
    let typeTypedef = {
      name: typedef.name,
      repr: typedef.repr,
      content: typeFull,
    };
    if (typedef.serialization === 'bytemuck') {
      return bytemuck(typeTypedef).value;
    }
    return ToolboxIdlTypeFull.typedef(typeTypedef);
  },
  generic: (
    self: ToolboxIdlTypeFlatGeneric,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull>,
    _typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull => {
    let typeFull = genericsBySymbol.get(self.symbol);
    if (typeFull === undefined) {
      throw new Error(`Could not resolve generic named: ${self.symbol}`);
    }
    return typeFull;
  },
  option: (
    self: ToolboxIdlTypeFlatOption,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull => {
    return ToolboxIdlTypeFull.option({
      prefix: self.prefix,
      content: hydrate(self.content, genericsBySymbol, typedefs),
    });
  },
  vec: (
    self: ToolboxIdlTypeFlatVec,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull => {
    return ToolboxIdlTypeFull.vec({
      prefix: self.prefix,
      items: hydrate(self.items, genericsBySymbol, typedefs),
    });
  },
  array: (
    self: ToolboxIdlTypeFlatArray,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull => {
    let length = hydrate(
      self.length,
      genericsBySymbol,
      typedefs,
    ).asConstLiteral();
    if (length === undefined) {
      throw new Error('Could not resolve array length as const literal');
    }
    return ToolboxIdlTypeFull.array({
      length,
      items: hydrate(self.items, genericsBySymbol, typedefs),
    });
  },
  string: (
    self: ToolboxIdlTypeFlatString,
    _genericsBySymbol: Map<string, ToolboxIdlTypeFull>,
    _typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull => {
    return ToolboxIdlTypeFull.string({
      prefix: self.prefix,
    });
  },
  struct: (
    self: ToolboxIdlTypeFlatStruct,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull => {
    return ToolboxIdlTypeFull.struct({
      fields: hydrateFields(self.fields, genericsBySymbol, typedefs),
    });
  },
  enum: (
    self: ToolboxIdlTypeFlatEnum,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull => {
    return ToolboxIdlTypeFull.enum({
      prefix: self.prefix,
      variants: self.variants.map((variant) => {
        return {
          name: variant.name,
          code: variant.code,
          fields: hydrateFields(variant.fields, genericsBySymbol, typedefs),
        };
      }),
    });
  },
  padded: (
    self: ToolboxIdlTypeFlatPadded,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull => {
    return ToolboxIdlTypeFull.padded({
      before: self.before,
      minSize: self.minSize,
      after: self.after,
      content: hydrate(self.content, genericsBySymbol, typedefs),
    });
  },
  const: (
    self: ToolboxIdlTypeFlatConst,
    _genericsBySymbol: Map<string, ToolboxIdlTypeFull>,
    _typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull => {
    return ToolboxIdlTypeFull.const({
      literal: self.literal,
    });
  },
  primitive: (
    self: ToolboxIdlTypePrimitive,
    _genericsBySymbol: Map<string, ToolboxIdlTypeFull>,
    _typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull => {
    return ToolboxIdlTypeFull.primitive(self);
  },
};

export function hydrateFields(
  typeFlatFields: ToolboxIdlTypeFlatFields,
  genericsBySymbol: Map<string, ToolboxIdlTypeFull>,
  typedefs: Map<string, ToolboxIdlTypedef>,
): ToolboxIdlTypeFullFields {
  return typeFlatFields.traverse(
    hydrateFieldsVisitor,
    genericsBySymbol,
    typedefs,
  );
}

let hydrateFieldsVisitor = {
  named: (
    self: ToolboxIdlTypeFlatFieldNamed[],
    genericsBySymbol: Map<string, ToolboxIdlTypeFull>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFullFields => {
    return ToolboxIdlTypeFullFields.named(
      self.map((field) => {
        return {
          name: field.name,
          content: hydrate(field.content, genericsBySymbol, typedefs),
        };
      }),
    );
  },
  unnamed: (
    self: ToolboxIdlTypeFlatFieldUnnamed[],
    genericsBySymbol: Map<string, ToolboxIdlTypeFull>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFullFields => {
    return ToolboxIdlTypeFullFields.unnamed(
      self.map((field, index) => {
        return {
          position: index,
          content: hydrate(field.content, genericsBySymbol, typedefs),
        };
      }),
    );
  },
};
